bchain/coins/nuls/nulsrpc.go

```diff
@@ -0,0 +1,616 @@
+		Address   string  `json:"address"`
+		Value     float64 `json:"value"`
+	} `json:"inputs"`
+	Outputs []struct {
+		Address  string `json:"address"`
+		Value    int64  `json:"value"`
+		LockTime int64  `json:"lockTime"`
+	} `json:"outputs"`
+}
+
+type CmdGetTxBytes struct {
+	Success bool `json:"success"`
+	Data    struct {
+		Value string `json:"value"`
+	} `json:"data"`
+}
+
+func (n *NulsRPC) GetChainInfo() (*bchain.ChainInfo, error) {
+	networkInfo := CmdGetNetworkInfo{}
+	error := n.Call("/api/network/info", &networkInfo)
+	if error != nil {
+		return nil, error
+	}
+
+	versionInfo := CmdGetVersionInfo{}
+	error = n.Call("/api/client/version", &versionInfo)
+	if error != nil {
+		return nil, error
+	}
+
+	chainInfo := &bchain.ChainInfo{
+		Chain:           "nuls",
+		Blocks:          networkInfo.Data.NetBestHeight,
+		Headers:         networkInfo.Data.NetBestHeight,
+		Bestblockhash:   "",
+		Difficulty:      networkInfo.Data.TimeOffset,
+		SizeOnDisk:      networkInfo.Data.LocalBestHeight,
+		Version:         versionInfo.Data.MyVersion,
+		Subversion:      versionInfo.Data.NewestVersion,
+		ProtocolVersion: strconv.Itoa(versionInfo.Data.NetworkVersion),
+		Timeoffset:      0,
+		Warnings:        versionInfo.Data.Infromation,
+	}
+	return chainInfo, nil
+}
+
+func (n *NulsRPC) GetBestBlockHash() (string, error) {
+	bestBlockHash := CmdGetBestBlockHash{}
+	error := n.Call("/api/block/newest/hash", &bestBlockHash)
+	if error != nil {
+		return "", error
+	}
+	return bestBlockHash.Data.Value, nil
+}
+
+func (n *NulsRPC) GetBestBlockHeight() (uint32, error) {
+	bestBlockHeight := CmdGetBestBlockHeight{}
+	error := n.Call("/api/block/newest/height", &bestBlockHeight)
+	if error != nil {
+		return 0, error
+	}
+	return bestBlockHeight.Data.Value, nil
+}
+
+func (n *NulsRPC) GetBlockHash(height uint32) (string, error) {
+	blockHeader := CmdGetBlockHeader{}
+	error := n.Call("/api/block/header/height/"+strconv.Itoa(int(height)), &blockHeader)
+	if error != nil {
+		return "", error
+	}
+
+	if !blockHeader.Success {
+		return "", bchain.ErrBlockNotFound
+	}
+
+	return blockHeader.Data.Hash, nil
+}
+func (n *NulsRPC) GetBlockHeader(hash string) (*bchain.BlockHeader, error) {
+	blockHeader := CmdGetBlockHeader{}
+	error := n.Call("/api/block/header/hash/"+hash, &blockHeader)
+	if error != nil {
+		return nil, error
+	}
+
+	if !blockHeader.Success {
+		return nil, bchain.ErrBlockNotFound
+	}
+
+	nexHash, _ := n.GetBlockHash(uint32(blockHeader.Data.Height + 1))
+
+	header := &bchain.BlockHeader{
+		Hash:          blockHeader.Data.Hash,
+		Prev:          blockHeader.Data.PreHash,
+		Next:          nexHash,
+		Height:        uint32(blockHeader.Data.Height),
+		Confirmations: blockHeader.Data.ConfirmCount,
+		Size:          blockHeader.Data.Size,
+		Time:          blockHeader.Data.Time,
+	}
+	return header, nil
+}
+
+func (n *NulsRPC) GetBlock(hash string, height uint32) (*bchain.Block, error) {
+	if hash == "" && height < 0 {
```

Comment:
You give me a chance to nitpick, uint32 is never less than zero :)

Reply:
Yes, I am wrong and have been modified.